### PR TITLE
Gracefully handle double responses

### PR DESCRIPTION
**Pull request by Raynos**
This is a WIP branch.

 - Improves errors in out_response
 - Add error printing to connection_base
 - Write tests for edge cases

TODO:

 - [x] more tests
 - [x] handle timeouts

r: @rf @kriskowal

### CODE REVIEW

File: node/connection_base.js [hunk 1]
```diff
@@ -331,6 +332,34 @@ TChannelConnectionBase.prototype.buildResponse = function buildResponse(req, opt
         done = true;
         self.onReqDone(req);
     }
+
+    function onError(err) {
+        self.onResError(err, req);
```
Comment (kriskowal) [inline node/connection_base.js]: onResponseError (we’ve been spelling req and res out on method names)

File: node/connection_base.js [hunk 2]
```diff
@@ -331,6 +332,34 @@ TChannelConnectionBase.prototype.buildResponse = function buildResponse(req, opt
...
+
+    self.logger.error('outgoing response has an error', {
+        err: err,
+        arg1: String(req.arg1),
+        ok: req.res.ok,
+        type: req.res.type,
+        bufArg2: arg2.slice(0, 50),
+        arg2: String(arg2).slice(0, 50),
+        bufArg3: arg3.slice(0, 50),
+        arg3: String(arg3).slice(0, 50)
```
Comment (kriskowal) [inline node/connection_base.js]: Perhaps do elided (…) hex dumps for both of these.

Comment (Raynos) [inline node/connection_base.js]: I thought about that; not sure what's the most useful to see

File: node/connection_base.js (hunk 1 above)
Comment (Raynos) [inline node/connection_base.js]: `onReqDone` right above it ;)

Comment (kriskowal) [inline node/connection_base.js]: Far more counter-examples. onReqDone is in the majority.

**Comment by Raynos** [comment]
@kriskowal @rf updated this PR

I've implemented more logging for error frames.

**Comment by Raynos** [comment]
Rebased and improved.

**warning:** PR is still not done; plenty of timeout edge cases to fix.

cc @rf @kriskowal

**Comment by Raynos** [comment]
@kriskowal @rf this is now ready for review.

**Comment by kriskowal** [comment]
:ok:

File: node/operations.js [hunk 3]
```diff
@@ -22,6 +22,8 @@
 
 var errors = require('./errors.js');
 
+var TOMBSTONE_TTL = 5000;
```
Comment (jcorbin) [inline node/operations.js]: should be something like 2*request ttl imo

File: node/operations.js [hunk 4]
```diff
@@ -197,6 +231,16 @@ function _onTimeoutCheck() {
 
     self._checkTimeout(self.requests.out, 'out');
     self._checkTimeout(self.requests.in, 'in');
+
+    var now = self.timers.now();
+    for (var i = 0; i < self.tombstones.out.length; i++) {
+        var tombstone = self.tombstones.out[i];
+        if (now >= tombstone.time + TOMBSTONE_TTL) {
+            self.tombstones.out.splice(i, 1);
```
Comment (jcorbin) [inline node/operations.js]: seems good'n'slow...

Comment (kriskowal) [inline node/operations.js]: https://github.com/kriskowal/pop-swap

Comment (kriskowal) [inline node/operations.js]: (Same big-O, but not as waseteful)

Comment (kriskowal) [inline node/operations.js]: Ignore previous. Tombstones should be in a map keyed by tracing.
